crates/datasources/src/excel/mod.rs

```diff
@@ -143,27 +163,69 @@ fn xlsx_sheet_value_to_record_batch(
 
     Ok(RecordBatch::try_new(Arc::new(schema), arrays)?)
 }
-pub async fn read_excel_impl(
-    path: &PathBuf,
-    sheet_name: Option<&str>,
+
+pub fn infer_schema(
+    r: &Range<calamine::Data>,
     has_header: bool,
     infer_schema_length: usize,
-) -> Result<datafusion::datasource::MemTable, Error> {
-    let mut workbook: Xlsx<_> = open_workbook(path)?;
-    let sheet = sheet_name.map(Cow::Borrowed).unwrap_or_else(|| {
-        let sheets = workbook.sheet_names();
-        let first = sheets.first().unwrap();
-        Cow::Owned(first.clone())
-    });
+) -> Result<Schema, ExcelError> {
+    let mut col_types: HashMap<&str, HashSet<DataType>> = HashMap::new();
+    let mut rows = r.rows();
+    let col_names: Vec<String> = rows
+        .next()
+        .unwrap()
+        .iter()
+        .enumerate()
+        .map(
+            |(i, c)| match (has_header, c.get_string().map(|s| s.to_string())) {
+                (true, Some(s)) => Ok(s),
+                (true, None) => Err(ExcelError::Load("failed to parse header".to_string())),
+                (false, _) => Ok(format!("col{}", i)),
+            },
+        )
+        .collect::<Result<_, _>>()?;
 
-    if let Ok(r) = workbook.worksheet_range(&sheet) {
-        let batch = xlsx_sheet_value_to_record_batch(r, has_header, infer_schema_length)?;
-        let schema_ref = batch.schema();
-        let partitions = vec![vec![batch]];
-        Ok(datafusion::datasource::MemTable::try_new(schema_ref, partitions).unwrap())
-    } else {
-        Err(Error::Load {
-            msg: "Failed to open .xlsx file.".to_owned(),
+    for row in rows.take(infer_schema_length) {
+        for (i, col_val) in row.iter().enumerate() {
+            let col_name = col_names.get(i).unwrap();
+            let col_type = infer_value_type(col_val).unwrap();
+            if col_type == DataType::Null {
+                continue;
+            }
+            let entry = col_types.entry(col_name).or_default();
+            entry.insert(col_type);
+        }
+    }
+
+    let fields: Vec<Field> = col_names
+        .iter()
+        .map(|col_name| {
+            let set = col_types.entry(col_name).or_insert_with(|| {
+                let mut set = HashSet::new();
+                set.insert(DataType::Utf8);
+                set
+            });
+
+            let mut dt_iter = set.iter().cloned();
+            let dt = dt_iter.next().unwrap_or(DataType::Utf8);
```

Comment:
can't we `into_iter` and avoid the mutable/extra declaration? 

is there a case where the datatype isn't string/utf8 here?

Reply:
> is there a case where the datatype isn't string/utf8 here?
added better err handling for this at the call site in `infer_value_type`, the `unwrap_or` here seems mostly to satisfy iter semantics returning an Option